src/sentry/integrations/utils/code_mapping.py

```diff
@@ -295,19 +284,34 @@ def _is_potential_match(self, src_file: str, frame_filename: FrameFilename) -> b
         Tries to see if the stacktrace without the root matches the file from the
         source code. Use existing code mappings to exclude some source files
         """
+
+        # l1 shoudl be longer
+        def _list_endswith(l1: list[str], l2: list[str]) -> bool:
+            if len(l2) > len(l1):
+                l1, l2 = l2, l1
+            l1_idx = len(l1) - 1
+            l2_idx = len(l2) - 1
+
+            while l2_idx >= 0:
+                if l2[l2_idx] != l1[l1_idx]:
+                    return False
+                l1_idx -= 1
+                l2_idx -= 1
+            return True
+
         # Exit early because we should not be processing source files for existing code maps
         if self._matches_existing_code_mappings(src_file):
             return False
 
         src_file_items = src_file.split("/")
-        frame_items = frame_filename.full_path.split("/")
-        if len(src_file_items) > len(frame_items):
-            return src_file.endswith(frame_filename.full_path)
-        else:
-            num_relevant_items = max(1, len(src_file_items) - 1)
-            relevant_items = frame_items[(len(frame_items) - num_relevant_items) :]
-            relevant_path = "/".join(relevant_items)
-            return src_file.endswith(relevant_path)
+        frame_items = frame_filename.normalized_path.split("/")
+
+        if len(src_file_items) > len(frame_items):  # Mono repos
+            return _list_endswith(src_file_items, frame_items)
+        elif len(frame_items) > len(src_file_items):  # Absolute paths
+            return _list_endswith(frame_items, src_file_items)
+        else:  # exact match
+            return src_file == frame_filename.normalized_path
```

Comment:
By matching based on path items rather than characters, we prevent certain cases where a stack path and source path are labeled as a match incorrectly. For example consider the following example:
* `stack_path = client.py`
* `source_path = app_client.py`
Our old algorithm would have labeled this as a match because technically the stack_path is a trailing substring of the source_path. 
This change prevents this from happening by changing the check from a _trailing substring_ to a _trailing file path_.